src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs

```diff
@@ -0,0 +1,153 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using System.Collections.Immutable;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.StatelessFuncIsnt
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis( CompilationStartAnalysisContext context ) {
+
+			Compilation compilation = context.Compilation;
+
+			context.RegisterSyntaxNodeAction(
+				ctx => {
+					if( ctx.Node is ObjectCreationExpressionSyntax expr ) {
+						AnalyzeInvocation( ctx, expr );
+					}
+				},
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private void AnalyzeInvocation(
+			SyntaxNodeAnalysisContext context,
+			ObjectCreationExpressionSyntax syntax
+		) {
+
+			TypeSyntax type = syntax.Type;
+			ISymbol symbol = context
+				.SemanticModel
+				.GetSymbolInfo( type ).Symbol;
+
+			if( symbol == null ) {
+				return;
+			}
+
+			if( !IsStatelessFunc( symbol ) ) {
+				return;
+			}
+
+			ExpressionSyntax argument = syntax.ArgumentList.Arguments[ 0 ].Expression;
+			SyntaxKind kind = argument.Kind();
+			switch( kind ) {
+
+				// this is the case when a method reference is used
+				// eg Func<string, int> func = int.Parse
+				case SyntaxKind.SimpleMemberAccessExpression:
+					if( IsStaticMemberAccess( context, argument ) ) {
+						return;
+					}
+
+					// non-static member access means that state could
+					// be used / held.
+					// TODO: Look for [Immutable] on the member's type
+					// to determine if the non-static member is safe
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda has parens
+				// eg () => 1, (x, y) => x + y
+				case SyntaxKind.ParenthesizedLambdaExpression:
+					if( !CheckForClosures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				// this is the case when the left hand side of the
+				// lambda does not have parens
+				// eg x => x + 1
+				case SyntaxKind.SimpleLambdaExpression:
+					if( !CheckForClosures( context, argument ) ) {
+						return;
+					}
+					break;
+
+				default:
+					// should we do something else here because constructors
+					// of D2L.StatelessFunc<T> and friends should always take
+					// a func?
+					return;
```

Comment:
I think this should probably be a diagnostic

```cs
Func<int> EvilFactory() {
  int x = 0;
  return () => {
    x = x + 1;
    return x;
  };
}

new StatelessFunc<T>( EvilFactory() )
```

Reply:
This case is an `InvocationExpressionSyntax`. Sounds like you're proposing we reject those to start?

Reply:
I'm proposing we reject everything by default :)

Reply:
Done